File: Homework/Hoang_Nguyen/lesson 1/303_range_sum_query_immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function (nums) {
  this.prefixSum = [];
  for (let i = 0; i < nums.length; i++) {
    this.prefixSum.push(nums[i] + (i ? this.prefixSum[i - 1] : 0));
  }
};

/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function (left, right) {
  return this.prefixSum[right] - (left ? this.prefixSum[left - 1] : 0);
};

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: Homework/Ngan_Luu/303-range-sum-query-immutable.py
#https://leetcode.com/problems/range-sum-query-immutable/
#Time:  O(n)
#Space: 0(n)
class NumArray:
    def __init__(self, nums: List[int]):
        self.prefixSum=nums
        for i in range(len(nums)-1):
            self.prefixSum[i+1]=self.prefixSum[i+1]+self.prefixSum[i]
    def sumRange(self, left: int, right: int) -> int:
        if left==0:
            return self.prefixSum[right]
        else:
            return self.prefixSum[right]-self.prefixSum[left-1]